Test failures on Python 3.13.0b1

I don't seem to be able to find a bug for tracking Python 3.13 support, so I hope it's fine if I file one.

```pytb
$ tox -e py313
.pkg: _optional_hooks> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_editable> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py313: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/pytest/.tox/.tmp/package/2/pytest-8.3.0.dev38+g93dd34e76.tar.gz
py313: commands[0]> pytest
============================= test session starts ==============================
platform linux -- Python 3.13.0b1, pytest-8.3.0.dev38+g93dd34e76, pluggy-1.5.0
cachedir: .tox/py313/.pytest_cache
rootdir: /tmp/pytest
configfile: pyproject.toml
testpaths: testing
plugins: hypothesis-6.102.1
collected 3668 items

[…]
=========================== short test summary info ============================
FAILED testing/code/test_excinfo.py::TestTraceback_f_g_h::test_traceback_recursion_index
FAILED testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_traceback_recursion
FAILED testing/code/test_excinfo.py::test_exception_repr_extraction_error_on_recursion
FAILED testing/code/test_source.py::test_getfslineno - assert 368 == -1
FAILED testing/test_collection.py::TestSession::test_collect_protocol_single_function
FAILED testing/test_collection.py::TestSession::test_collect_custom_nodes_multi_id
FAILED testing/test_collection.py::TestSession::test_collect_subdir_event_ordering
FAILED testing/test_collection.py::TestSession::test_collect_two_commandline_args
FAILED testing/test_doctest.py::TestDoctests::test_doctest_unexpected_exception
FAILED testing/test_doctest.py::TestDoctests::test_doctest_linedata_on_property
FAILED testing/test_legacypath.py::test_testdir_makefile_ext_none_raises_type_error
XPASS testing/_py/test_local.py::TestLocalPath::test_make_numbered_dir_multiprocess_safe - #11603
= 11 failed, 3535 passed, 110 skipped, 11 xfailed, 1 xpassed in 138.50s (0:02:18) =
py313: exit 1 (139.10 seconds) /tmp/pytest> pytest pid=65870
  py313: FAIL code 1 (145.84=setup[6.73]+cmd[139.10] seconds)
  evaluation failed :( (146.00 seconds)
```

This is on Gentoo Linux amd64, with 93dd34e76d9c687d1c249fe8cf94bdf46813f783.

Full log: 
[py313.txt](https://github.com/pytest-dev/pytest/files/15310368/py313.txt)

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -55,6 +55,7 @@ jobs:
           "windows-py310",
           "windows-py311",
           "windows-py312",
+          "windows-py313",
 
           "ubuntu-py38",
           "ubuntu-py38-pluggy",
@@ -63,12 +64,14 @@ jobs:
           "ubuntu-py310",
           "ubuntu-py311",
           "ubuntu-py312",
+          "ubuntu-py313",
           "ubuntu-pypy3",
 
           "macos-py38",
           "macos-py39",
           "macos-py310",
           "macos-py312",
+          "macos-py313",
 
           "doctesting",
           "plugins",
@@ -97,9 +100,13 @@ jobs:
             os: windows-latest
             tox_env: "py311"
           - name: "windows-py312"
-            python: "3.12-dev"
+            python: "3.12"
             os: windows-latest
             tox_env: "py312"
+          - name: "windows-py313"
+            python: "3.13-dev"
+            os: windows-latest
+            tox_env: "py313"
 
           - name: "ubuntu-py38"
             python: "3.8"
@@ -128,9 +135,14 @@ jobs:
             tox_env: "py311"
             use_coverage: true
           - name: "ubuntu-py312"
-            python: "3.12-dev"
+            python: "3.12"
             os: ubuntu-latest
             tox_env: "py312"
+            use_coverage: true
+          - name: "ubuntu-py313"
+            python: "3.13-dev"
+            os: ubuntu-latest
+            tox_env: "py313"
             use_coverage: true
           - name: "ubuntu-pypy3"
             python: "pypy-3.8"
@@ -151,9 +163,13 @@ jobs:
             os: macos-latest
             tox_env: "py310-xdist"
           - name: "macos-py312"
-            python: "3.12-dev"
+            python: "3.12"
             os: macos-latest
             tox_env: "py312-xdist"
+          - name: "macos-py313"
+            python: "3.13-dev"
+            os: macos-latest
+            tox_env: "py313-xdist"
 
           - name: "plugins"
             python: "3.12"

--- a/changelog/12334.improvement.rst
+++ b/changelog/12334.improvement.rst
@@ -1,0 +1,1 @@
+Support for Python 3.13 (beta1 at the time of writing).

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -31,6 +31,7 @@ classifiers = [
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13",
     "Topic :: Software Development :: Libraries",
     "Topic :: Software Development :: Testing",
     "Topic :: Utilities",

--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -424,15 +424,14 @@ class Traceback(List[TracebackEntry]):
             # which generates code objects that have hash/value equality
             # XXX needs a test
             key = entry.frame.code.path, id(entry.frame.code.raw), entry.lineno
-            # print "checking for recursion at", key
             values = cache.setdefault(key, [])
+            # Since Python 3.13 f_locals is a proxy, freeze it.
+            loc = dict(entry.frame.f_locals)
             if values:
-                f = entry.frame
-                loc = f.f_locals
                 for otherloc in values:
                     if otherloc == loc:
                         return i
-            values.append(entry.frame.f_locals)
+            values.append(loc)
         return None
 
 

--- a/src/_pytest/pytester.py
+++ b/src/_pytest/pytester.py
@@ -289,7 +289,8 @@ class HookRecorder:
         __tracebackhide__ = True
         i = 0
         entries = list(entries)
-        backlocals = sys._getframe(1).f_locals
+        # Since Python 3.13, f_locals is not a dict, but eval requires a dict.
+        backlocals = dict(sys._getframe(1).f_locals)
         while entries:
             name, check = entries.pop(0)
             for ind, call in enumerate(self.calls[i:]):
@@ -759,6 +760,9 @@ class Pytester:
         encoding: str = "utf-8",
     ) -> Path:
         items = list(files.items())
+
+        if ext is None:
+            raise TypeError("ext must not be None")
 
         if ext and not ext.startswith("."):
             raise ValueError(

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1,6 +1,7 @@
 # mypy: allow-untyped-defs
 from __future__ import annotations
 
+import fnmatch
 import importlib
 import io
 import operator
@@ -237,7 +238,7 @@ class TestTraceback_f_g_h:
                 n += 1
             f(n)
 
-        excinfo = pytest.raises(RuntimeError, f, 8)
+        excinfo = pytest.raises(RecursionError, f, 8)
         traceback = excinfo.traceback
         recindex = traceback.recursionindex()
         assert recindex == 3
@@ -373,7 +374,10 @@ def test_excinfo_no_sourcecode():
     except ValueError:
         excinfo = _pytest._code.ExceptionInfo.from_current()
     s = str(excinfo.traceback[-1])
-    assert s == "  File '<string>':1 in <module>\n  ???\n"
+    # TODO: Since Python 3.13b1 under pytest-xdist, the * is `import
+    # sys;exec(eval(sys.stdin.readline()))` (execnet bootstrap code)
+    # instead of `???` like before. Is this OK?
+    fnmatch.fnmatch(s, "  File '<string>':1 in <module>\n  *\n")
 
 
 def test_excinfo_no_python_sourcecode(tmp_path: Path) -> None:

--- a/testing/code/test_source.py
+++ b/testing/code/test_source.py
@@ -370,7 +370,11 @@ def test_getfslineno() -> None:
         pass
 
     B.__name__ = B.__qualname__ = "B2"
-    assert getfslineno(B)[1] == -1
+    # Since Python 3.13 this started working.
+    if sys.version_info >= (3, 13):
+        assert getfslineno(B)[1] != -1
+    else:
+        assert getfslineno(B)[1] == -1
 
 
 def test_code_of_object_instance_with_call() -> None:

--- a/testing/test_cacheprovider.py
+++ b/testing/test_cacheprovider.py
@@ -194,7 +194,7 @@ class TestNewAPI:
         assert pytester.path.joinpath("custom_cache_dir").is_dir()
 
 
-@pytest.mark.parametrize("env", ((), ("TOX_ENV_DIR", "/tox_env_dir")))
+@pytest.mark.parametrize("env", ((), ("TOX_ENV_DIR", "mydir/tox-env")))
 def test_cache_reportheader(
     env: Sequence[str], pytester: Pytester, monkeypatch: MonkeyPatch
 ) -> None:

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -1,6 +1,7 @@
 # mypy: allow-untyped-defs
 import inspect
 from pathlib import Path
+import sys
 import textwrap
 from typing import Callable
 from typing import Optional
@@ -223,6 +224,7 @@ class TestDoctests:
                 "Traceback (most recent call last):",
                 '  File "*/doctest.py", line *, in __run',
                 "    *",
+                *((" *^^^^*", " *", " *") if sys.version_info >= (3, 13) else ()),
                 '  File "<doctest test_doctest_unexpected_exception.txt[1]>", line 1, in <module>',
                 "ZeroDivisionError: division by zero",
                 "*/test_doctest_unexpected_exception.txt:2: UnexpectedException",
@@ -379,7 +381,7 @@ class TestDoctests:
                 "*= FAILURES =*",
                 "*_ [[]doctest[]] test_doctest_linedata_on_property.Sample.some_property _*",
                 "004 ",
-                "005         >>> Sample().some_property",
+                "005 *>>> Sample().some_property",
                 "Expected:",
                 "    'another thing'",
                 "Got:",

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -1,7 +1,6 @@
 # mypy: allow-untyped-defs
 import inspect
 from pathlib import Path
-import sys
 import textwrap
 from typing import Callable
 from typing import Optional
@@ -224,11 +223,6 @@ class TestDoctests:
                 "Traceback (most recent call last):",
                 '  File "*/doctest.py", line *, in __run',
                 "    *",
-                *(
-                    (" *^^^^*",)
-                    if (3, 11, 0, "beta", 4) > sys.version_info >= (3, 11)
-                    else ()
-                ),
                 '  File "<doctest test_doctest_unexpected_exception.txt[1]>", line 1, in <module>',
                 "ZeroDivisionError: division by zero",
                 "*/test_doctest_unexpected_exception.txt:2: UnexpectedException",

--- a/testing/test_main.py
+++ b/testing/test_main.py
@@ -3,7 +3,6 @@ import argparse
 import os
 from pathlib import Path
 import re
-import sys
 from typing import Optional
 
 from _pytest.config import ExitCode
@@ -45,32 +44,18 @@ def test_wrap_session_notify_exception(ret_exc, pytester: Pytester) -> None:
         assert result.ret == ExitCode.INTERNAL_ERROR
     assert result.stdout.lines[0] == "INTERNALERROR> Traceback (most recent call last):"
 
-    end_lines = (
-        result.stdout.lines[-4:]
-        if (3, 11, 0, "beta", 4) > sys.version_info >= (3, 11)
-        else result.stdout.lines[-3:]
-    )
+    end_lines = result.stdout.lines[-3:]
 
     if exc == SystemExit:
         assert end_lines == [
             f'INTERNALERROR>   File "{c1}", line 4, in pytest_sessionstart',
             'INTERNALERROR>     raise SystemExit("boom")',
-            *(
-                ("INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^",)
-                if (3, 11, 0, "beta", 4) > sys.version_info >= (3, 11)
-                else ()
-            ),
             "INTERNALERROR> SystemExit: boom",
         ]
     else:
         assert end_lines == [
             f'INTERNALERROR>   File "{c1}", line 4, in pytest_sessionstart',
             'INTERNALERROR>     raise ValueError("boom")',
-            *(
-                ("INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^",)
-                if (3, 11, 0, "beta", 4) > sys.version_info >= (3, 11)
-                else ()
-            ),
             "INTERNALERROR> ValueError: boom",
         ]
     if returncode is False:

--- a/tox.ini
+++ b/tox.ini
@@ -9,6 +9,7 @@ envlist =
     py310
     py311
     py312
+    py313
     pypy3
     py38-{pexpect,xdist,unittestextras,numpy,pluggymain,pylib}
     doctesting

